pkg/cli/root.go

```diff
@@ -12,89 +12,83 @@
+	vpr := viper.New()
 
-func init() {
-	cobra.OnInitialize(initCfg)
-	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.jvsctl/config.yaml)")
-	rootCmd.PersistentFlags().String("server", "", "overwrite the JVS server address")
-	rootCmd.PersistentFlags().Bool("insecure", false, "use insecure connection to JVS server")
-	viper.BindPFlag("server", rootCmd.PersistentFlags().Lookup("server"))     //nolint // not expect err
-	viper.BindPFlag("insecure", rootCmd.PersistentFlags().Lookup("insecure")) //nolint // not expect err
+	opts := &rootCmdOptions{}
 
-	rootCmd.AddCommand(tokenCmd)
-}
+	cmd := &cobra.Command{
+		Use:   "jvsctl",
+		Short: "jvsctl facilitates the justification verification flow provided by abcxyz/jvs",
 
-func initCfg() {
-	if cfgFile != "" {
-		// Use config file from the flag.
-		viper.SetConfigFile(cfgFile)
-	} else {
-		// Try use the default config file.
-		home, err := os.UserHomeDir()
-		cobra.CheckErr(err)
-
-		viper.AddConfigPath(filepath.Join(home, ".jvsctl"))
-		viper.SetConfigType("yaml")
-		viper.SetConfigName("config")
+		// Load the configuration before each child command. The actual "compiled"
+		// configuration is passed in to the newChildCmd() function, but this sets
+		// the values of that config.
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return loadConfig(vpr, opts.configPath, cfg)
+		},
 	}
 
-	// Also load from env vars.
-	viper.SetEnvPrefix("JVSCTL")
-	viper.AutomaticEnv()
+	// Flags
+	cmd.PersistentFlags().StringVarP(&opts.configPath, "config", "c", "",
+		"path to a file on disk for the jvs configuration")
+	cmd.PersistentFlags().StringVar(&cfg.Server, "server", "127.0.0.1:8080", "IP or DNS address to the JVS server")
+	cmd.PersistentFlags().BoolVar(&cfg.Insecure, "insecure", false, "allow an insecure connection to the JVS server")
 
-	if err := viper.ReadInConfig(); err != nil {
-		// It's ok if the config file is not found because
-		// the values could be filled by env vars or flags.
-		if !errors.As(err, &viper.ConfigFileNotFoundError{}) {
-			cobra.CheckErr(err)
-			return
-		}
+	// Subcommands
+	cmd.AddCommand(newTokenCmd(cfg))
+
+	return cmd
+}
+
+// loadConfig loads the configuration from the given path into the provided
+// configuration struct.
+func loadConfig(vpr *viper.Viper, pth string, cfg *config.CLIConfig) error {
+	if pth == "" {
+		pth = defaultConfigPath
 	}
 
-	if err := viper.Unmarshal(&cfg); err != nil {
-		cobra.CheckErr(err)
-		return
+	vpr.SetConfigType("yaml")
+	vpr.SetEnvPrefix("JVSCTL")
+	vpr.AutomaticEnv()
+	vpr.SetConfigFile(pth)
+
+	if err := vpr.ReadInConfig(); err != nil {
+		// Don't throw an error for failing to find the default config path since
+		// it's not required. However, if the user specified a config path that
+		// doesn't exist, that's an error.
+		if !errors.Is(err, &viper.ConfigFileNotFoundError{}) && !errors.Is(err, os.ErrNotExist) {
```

Comment:
I thought viper returns ConfigFileNotFoundError if os.ErrNotExist. Are you doing this for safety?

Reply:
Nope. I found it returns `ConfigFileNotFoundError` when given a _file_ path that does not exist, but returns `ErrNotExist` when given a _directory_ path that does not exist. `ConfigFileNotFoundError` only seemed to throw when the file had an extension.